### PR TITLE
refactor: reduce command boilerplate and centralize file watching

### PR DESCRIPTION
**Pull request by JanZachmann**
## Summary

- Add `parse_payload()` helper to deduplicate JSON parsing in `from_direct_method()` and `from_desired_property()` (fixes copy-paste bug in `set_wait_online_timeout` error message)
- Unify `DirModified`/`FileCreated`/`FileModified` command variants into a single `FsEvent(FsEventCommand)` with `FsEventKind` discriminator; remove `PathCommand` struct
- Drop unused `instant` field from `IntervalCommand`
- Split `feature/mod.rs` into `mod.rs` (Feature trait) + `command.rs` (Command types) + `fs_watcher.rs` (FsWatcher)
- Replace `notify`/`notify-debouncer-full` with a centralized `FsWatcher` backed by a single `inotify` instance — no locks, per-watch trailing-edge debounce, native `ONESHOT` support for file-creation watches
- Add 11 new unit tests covering FsWatcher (event delivery, debounce, oneshot, race condition, noop), DirModified command dispatch, and `feature_id()` routing
- Bump version to 0.42.5
- Apply `cargo fmt` import grouping (`imports_granularity = Crate`)

## Reason

The `Command` enum infrastructure had significant boilerplate: repetitive `serde_json::from_value` + error handling in parsing functions, three near-identical file-event variants with three near-identical stream factory functions, and per-feature `notify-debouncer-full` debouncers with no shared state. This refactor reduces touch-points when adding new commands and consolidates file watching into one efficient inotify instance.

### CODE REVIEW

File: src/twin/feature/fs_watcher.rs [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+                    .context("add_watch: FileCreated path has no parent")?;
+                let mut mask = WatchMask::CREATE;
+                if oneshot {
+                    mask |= WatchMask::ONESHOT;
+                }
+                (parent, mask)
+            }
+            FsEventKind::FileModified => (path, WatchMask::CLOSE_WRITE),
+            FsEventKind::DirModified => (path, WatchMask::CREATE | WatchMask::DELETE),
+        };
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in e711f3b: removed `WatchMask::ONESHOT` from `add_watch`. Oneshot is now enforced in user-space — the event loop removes the `watch_info` entry and the kernel watch after dispatching a matching filename event. Regression test added in cf08a31 (`file_created_oneshot_ignores_unrelated_files`).

File: src/twin/feature/fs_watcher.rs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+                let _ = tx.try_send(CommandRequest {
+                    command: Command::FsEvent(FsEventCommand {
+                        kind: FsEventKind::FileCreated,
+                        feature_id: info.feature_id,
+                        path: info.path.clone(),
+                    }),
+                    reply: None,
+                });
+            }
+        }
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in e711f3b: `try_send` errors are now logged. Oneshot `watch_info` entries for pre-existing files are removed via `retain()` after the immediate dispatch loop, preventing stale state. Regression test added in cf08a31 (`file_created_race_no_duplicate`).

File: src/twin/feature/fs_watcher.rs [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+
+        tokio::spawn(async move {
+            let mut buffer = [0; 1024];
+            let mut stream = inotify
+                .into_event_stream(&mut buffer)
+                .expect("into_stream: failed to create event stream");
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in e711f3b: replaced `expect()` with `match` + `error!()` + `return`. The spawned task now logs the error and exits gracefully instead of panicking.

File: project-context.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  - `modemmanager` (optional, behind `modem_info` feature) — modem status via D-Bus
+  - `sd-notify` — systemd readiness notification
+  - `sysinfo` — disk, component, and system information
+
+## 3. Key Entry Points & Files
+
+- `src/main.rs` — binary entry point; sets up logging, delegates to `Twin::run()`
+- `src/twin/mod.rs` — core event loop: creates IoT Hub client, initializes all features, dispatches commands from direct methods / desired properties / file watches / intervals
+- `src/twin/feature/mod.rs` — `Feature` trait definition, `Command` enum (all dispatchable operations), `CommandRequest` types
+- `src/web_service.rs` — actix-web HTTP server exposing publish channels (`/publish/v1/{channel}`) and status endpoints; manages publish endpoints and retry logic
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply project-context.md:50]: `project-context.md` is not committed to the repo (listed in `.gitignore`). It's a local working document, not part of the PR.

File: project-context.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+- `systemd/` — unit files: `omnect-device-service.service`, `.socket`, `.timer`, plus update-validation-observer
+- `sudo/` — sudoers drop-in files granting the service permission for specific privileged commands
+- `polkit/` — polkit rules for D-Bus permissions
+- `testfiles/` — test fixtures (positive/negative cases, Wi-Fi commissioning configs)
+
+## 4. Repository-Specific Constraints
+
+- **Mutually exclusive feature flags:** Exactly one of `bootloader_grub`, `bootloader_uboot`, or `mock` must be active. The `mock` feature stubs out hardware/systemd interactions for testing.
+- **Feature trait pattern:** Every device capability implements the `Feature` trait (in its own `src/twin/*.rs` module). Adding a new feature means: (1) implement `Feature`, (2) add a `Command` variant, (3) register in `Twin::new()` feature map.
+- **Command dispatch:** All operations flow through the `Command` enum in `feature/mod.rs`. Direct methods, desired properties, file-system events, and intervals all produce `Command` values that get routed to the owning feature via `TypeId`.
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply project-context.md]: Same as above — `project-context.md` is in `.gitignore`, not part of the PR.

File: project-context.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (JanZachmann) [reply project-context.md:50]: Correction: this file IS in the PR. Fixed in 02bab90 — updated dependency reference from `notify`/`notify-debouncer-full` to `inotify`, and updated key file paths to reflect the `command.rs`/`fs_watcher.rs` split.

File: project-context.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (JanZachmann) [reply project-context.md]: Correction: this file IS in the PR. Fixed in 02bab90 — updated command dispatch description to reference `src/twin/feature/command.rs` and `fs_watcher.rs`.

File: src/twin/feature/fs_watcher.rs [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+
+        self.watch_info.insert(
+            wd.get_watch_descriptor_id(),
+            WatchInfo {
+                feature_id: TypeId::of::<T>(),
+                path: path.to_path_buf(),
+                kind,
+                oneshot,
+            },
+        );
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in 8de79fc: changed `watch_info` from `HashMap<c_int, WatchInfo>` to `HashMap<c_int, (WatchDescriptor, Vec<WatchInfo>)>`. Multiple targets per wd_id are now supported — `add_watch` appends to the vec via the entry API. The event loop matches by filename for FileCreated and dispatches/removes per-entry. Kernel watch is only removed when the vec empties. Regression test added (`file_created_multiple_targets_same_dir`).

File: src/twin/feature/fs_watcher.rs [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+                }
+            }
+        }
+
+        // Clean up oneshot watch_info for files that were already dispatched
+        self.watch_info.retain(|_, info| {
+            !(info.oneshot
+                && info.kind == FsEventKind::FileCreated
+                && matches!(info.path.try_exists(), Ok(true)))
+        });
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in 8de79fc: the `WatchDescriptor` is now stored alongside the vec in `HashMap<c_int, (WatchDescriptor, Vec<WatchInfo>)>`. After `retain()` removes satisfied oneshot entries, empty map entries are collected and their kernel watches removed via `watches.remove(wd)`. Same cleanup happens in the event loop when a oneshot fires and empties the vec.

File: src/twin/feature/fs_watcher.rs [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+
+        let (watch_path, mask) = match kind {
+            FsEventKind::FileCreated => {
+                let parent = path
+                    .parent()
+                    .context("add_watch: FileCreated path has no parent")?;
+                // No WatchMask::ONESHOT here — any CREATE in the parent dir would
+                // consume it before our target filename is seen. Oneshot semantics
+                // are enforced in user-space by the event loop instead.
+                (parent, WatchMask::CREATE)
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in 0c9b0fb: added `WatchMask::MOVED_TO` to the FileCreated mask. Files appearing via atomic rename/move are now detected alongside direct creation.

File: src/twin/consent.rs [hunk 9]
```diff
@@ -119,6 +110,20 @@ impl DeviceUpdateConsent {
     const USER_CONSENT_VERSION: u8 = 1;
     const ID: &'static str = "device_update_consent";
 
+    pub fn new(fs_watcher: &mut FsWatcher) -> Result<Self> {
+        for path in [
+            request_consent_path!(),
+            history_consent_path!(),
+            general_consent_path!(),
+        ] {
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/consent.rs]: Fixed in 4893d44: removed `general_consent_path!()` from the watch list. This was incorrectly added during the FsWatcher migration — the original code only watched `request_consent` and `history_consent`. `update_general_consent()` already writes + reports, so watching the file produces a delayed duplicate.

File: src/twin/consent.rs [hunk 10]
```diff
@@ -119,6 +110,16 @@ impl DeviceUpdateConsent {
     const USER_CONSENT_VERSION: u8 = 1;
     const ID: &'static str = "device_update_consent";
 
+    pub fn new(fs_watcher: &mut FsWatcher) -> Result<Self> {
+        for path in [request_consent_path!(), history_consent_path!()] {
+            fs_watcher.add_watch::<DeviceUpdateConsent>(&path, FsEventKind::FileModified, false)?;
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/consent.rs]: Fixed in 7c7fdb9: changed `handle_request` to silently drop commands for disabled features when there's no reply channel (FsEvents, intervals). Direct methods and web service requests still fail loudly. This is centralized — no per-feature `is_enabled` checks needed in constructors.

File: src/twin/system_info.rs [hunk 11]
```diff
@@ -215,6 +206,7 @@ impl SystemInfo {
             Some(Self::boot_time()?)
         } else {
             debug!("new: start timesync watcher since not synced yet");
+            fs_watcher.add_watch::<SystemInfo>(&TIMESYNC_FILE, FsEventKind::FileCreated, true)?;
             None
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/system_info.rs:211]: Fixed in 7c7fdb9: changed `handle_request` to silently drop commands for disabled features when there's no reply channel (FsEvents, intervals). Direct methods and web service requests still fail loudly. This is centralized — no per-feature `is_enabled` checks needed in constructors.

File: src/twin/factory_reset.rs [hunk 12]
```diff
@@ -187,7 +180,13 @@ impl FactoryReset {
     const FACTORY_RESET_VERSION: u8 = 3;
     const ID: &'static str = "factory_reset";
 
-    pub fn new() -> Result<Self> {
+    pub fn new(fs_watcher: &mut FsWatcher) -> Result<Self> {
+        fs_watcher.add_watch::<FactoryReset>(
+            Path::new(&custom_config_dir_path!()),
+            FsEventKind::DirModified,
+            false,
+        )?;
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/factory_reset.rs]: Fixed in 7c7fdb9: changed `handle_request` to silently drop commands for disabled features when there's no reply channel (FsEvents, intervals). Direct methods and web service requests still fail loudly. This is centralized — no per-feature `is_enabled` checks needed in constructors.

File: src/twin/feature/fs_watcher.rs [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+                                    let _ = tx.send(CommandRequest {
+                                        command: Command::FsEvent(FsEventCommand {
+                                            kind: info.kind.clone(),
+                                            feature_id: info.feature_id,
+                                            path: info.path.clone(),
+                                        }),
+                                        reply: None,
+                                    }).await;
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in 806d5c3: `add_watch` now enforces `oneshot=true` for `FileCreated` via `ensure!`. A non-oneshot FileCreated watch doesn't make semantic sense (file creation is a one-time event), and the debounce path dispatches without filename filtering, so allowing it would cause spurious events.

File: src/twin/feature/fs_watcher.rs [hunk 14]
```diff
@@ -0,0 +1,669 @@
...
+            .collect();
+        for wd_id in empty_wds {
+            if let Some((wd, _)) = self.watch_info.remove(&wd_id)
+                && let Some(watches) = &mut self.watches
+            {
+                let _ = watches.remove(wd);
+            }
+        }
+
+        tokio::spawn(async move {
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in 6c83cd7: moved the entire race-condition handling (dispatch + cleanup) into the spawned async task. Uses `tx.send().await` for guaranteed delivery. Oneshot entries are only cleaned up after successful send.

File: Cargo.toml [hunk 15]
```diff
@@ -65,24 +64,25 @@ tokio = { version = "1", default-features = false }
 tokio-stream = { version = "0.1", default-features = false, features = [
   "time",
 ] }
+tokio-util = { version = "0.7", default-features = false }
```
Comment (JoergZeidler) [inline Cargo.toml:67]: **[Major]** `tokio-util` is declared with `default-features = false` but no features are listed. `CancellationToken` lives in `tokio_util::sync` and requires the `sync` feature. This compiles today only because a transitive dependency (likely `tokio-stream`) activates `sync` via feature unification — not because it is explicitly declared here. A future dependency update could silently break the build.

```toml
tokio-util = { version = "0.7", default-features = false, features = ["sync"] }
```

File: src/twin/feature/fs_watcher.rs [hunk 16]
```diff
@@ -0,0 +1,1410 @@
...
+            FsEventKind::FileModified,
+        )
+    }
+
+    /// Watch a directory for entry additions and deletions (coalesced into
+    /// one `FsEventKind::DirModified` command per debounce window).
+    pub fn watch_dir_modified<T: 'static>(&mut self, path: &Path) -> Result<()> {
+        self.register_modified::<T>(
+            path,
+            WatchMask::CREATE | WatchMask::DELETE | WatchMask::MASK_ADD,
```
Comment (JoergZeidler) [inline src/twin/feature/fs_watcher.rs]: **[Major]** `watch_dir_modified` only subscribes to `CREATE | DELETE`. An atomic rename into or out of the watched directory (`MOVED_TO` / `MOVED_FROM`) produces neither event, so config files written via `mv` / rename-swap (e.g. `swupdate`, many config management tools) are silently missed.

Consider adding:
```rust
WatchMask::CREATE | WatchMask::DELETE | WatchMask::MOVED_TO | WatchMask::MOVED_FROM | WatchMask::MASK_ADD
```
The doc comment on line 105 also claims "additions and deletions" but should say "additions, deletions and renames".

File: src/twin/system_info.rs [hunk 17]
```diff
@@ -215,6 +207,7 @@ impl SystemInfo {
             Some(Self::boot_time()?)
         } else {
             debug!("new: start timesync watcher since not synced yet");
+            fs_watcher.watch_file_created_oneshot::<SystemInfo>(&TIMESYNC_FILE)?;
```
Comment (JoergZeidler) [inline src/twin/system_info.rs:218]: **[Major]** `watch_file_created_oneshot` installs an inotify watch on the **parent directory** synchronously (`inotify::Watches::add`). If `/run/systemd/timesync/` does not exist at the time `SystemInfo::new()` runs, the `add()` call returns an error and `?` propagates it, aborting service startup.

The old `file_created_stream()` used `try_exists()` and was resilient to a missing parent dir. Consider checking whether the parent exists first and falling back gracefully, or ensure the systemd unit has an `After=systemd-tmpfiles-setup.service` dependency that guarantees `/run/systemd/timesync/` is present.

File: src/twin/mod.rs [hunk 18]
```diff
@@ -257,11 +270,22 @@ impl Twin {
...
-            feature.name()
-        );
+        if !feature.is_enabled() {
+            // FsEvents and intervals have no reply channel — drop them for
+            // disabled features. Direct methods and web service requests carry a
+            // reply channel and must receive the disabled reason; bailing here
+            // instead would drop the responder (surfacing to the caller as a
+            // generic transport error) and abort the run loop, restarting the
+            // whole service over a single request to a suppressed feature.
+            let msg = format!("handle_request: feature is disabled {}", feature.name());
```
Comment (JoergZeidler) [inline src/twin/mod.rs:280]: **[Minor]** `info!` here fires for every `FsEvent` and `Tick` dispatched to a suppressed feature (which can be very frequent — on every file change or interval). This will produce significant log spam in production if a feature is disabled via `SUPPRESS_*`.

Consider `debug!` for the no-reply path (FsEvent/Tick) and keep `info!` only when there is a `reply` channel (i.e., an explicit caller is waiting):
```rust
if reply.is_some() {
    info!("{msg}");
} else {
    debug!("{msg}");
}
```

File: src/twin/mod.rs [hunk 19]
```diff
@@ -337,10 +387,18 @@ impl Twin {
...
-                } else {
-                    None
+                if !f.is_enabled() {
+                    return None;
+                }
+                match f.command_request_stream(self.cancel.child_token()) {
+                    Ok(stream) => stream,
+                    Err(e) => {
+                        error!(
+                            "feature_command_request_streams: {} failed: {e:#}",
```
Comment (JoergZeidler) [inline src/twin/mod.rs:410]: **[Minor]** Changed from `.unwrap()` (panic → loud crash → systemd restart) to `log error + None` (silent degradation). This means a misconfigured feature stream (e.g., inotify setup error at startup) will not fail the service — it will silently produce no events. Since `FsWatcher` errors are already surfaced via `Twin::new()`, this is unlikely to hide real issues, but the behavioral change (panic → silent skip) is worth documenting with a comment.

File: src/twin/feature/command.rs [hunk 20]
```diff
@@ -0,0 +1,904 @@
...
+    silence: Duration,
+    cancel: CancellationToken,
+) -> CommandRequestStream
+where
+    F: Future<Output = Result<S>> + Send + 'static,
+    S: Stream<Item = I> + Send + 'static,
+    I: Send,
+    T: 'static,
+{
+    // Capacity 1: debounce guarantees at most one pending request
```
Comment (JoergZeidler) [inline src/twin/feature/command.rs]: **[Nit]** The comment "Capacity 1: debounce guarantees at most one pending request" overstates the guarantee. With capacity 1, `tx.send(req).await` can block if the consumer is slow. While the send is awaited, the `cancel.cancelled()` arm in the `select!` loop above cannot be polled — stretching shutdown latency by up to one channel drain cycle. The capacity is fine, but the comment should say "backpressure: at most one queued event" rather than implying a hard ordering guarantee.

File: src/twin/feature/command.rs [hunk 21]
```diff
@@ -0,0 +1,904 @@
...
+
+                _ = cancel.cancelled() => return,
+
+                item = stream.next() => match item {
+                    Some(_) => {
+                        debug!("debounced_command_stream: event, (re)arming debounce");
+                        deadline = Some(tokio::time::Instant::now() + silence);
+                    }
+                    None => {
+                        error!("debounced_command_stream: source stream ended, feature will no longer receive events");
```
Comment (JoergZeidler) [inline src/twin/feature/command.rs:258]: **[Missing test]** This `None` branch — source stream ending unexpectedly after previously delivering events — is not covered by any test. The `debounced_command_stream_source_init_failure_test` only covers init-time failure. A test that sends events then drops the sender would exercise this recovery path, which is critical for the `Network` feature if the D-Bus signal stream dies.

File: src/twin/feature/fs_watcher.rs [hunk 22]
```diff
@@ -0,0 +1,1410 @@
...
+    // Q_OVERFLOW: the kernel dropped events we cannot reconstruct. Bail out
+    // so systemd restarts and rebuilds a consistent view rather than
+    // silently missing file-state transitions.
+    if event.mask.contains(EventMask::Q_OVERFLOW) {
+        error!("FsWatcher: inotify queue overflow, exiting to let systemd restart");
+        return ControlFlow::Break(());
+    }
+
+    let wd = event.wd.clone();
+    let has_modified = modified.get(&wd).is_some_and(|l| !l.is_empty());
```
Comment (JoergZeidler) [inline src/twin/feature/fs_watcher.rs:356]: **[Missing test]** The `Q_OVERFLOW` exit path is untested. Since `handle_inotify_event` is a free `async fn` with all state passed as parameters, a unit test can synthesize a `Q_OVERFLOW` event without needing to actually overflow the kernel queue. This would catch any future refactoring that accidentally removes the overflow guard.

File: Cargo.toml (hunk 15 above)
Comment (JanZachmann) [reply Cargo.toml:67]: Not changed — this finding is based on an older \`tokio-util\`. In 0.7.18 there is no \`sync\` feature anymore: adding \`features = ["sync"]\` fails to resolve (\`tokio-util does not have that feature\`). In this version \`pub mod sync;\` in \`lib.rs\` is compiled unconditionally (no \`#[cfg(feature)]\` gate) and \`CancellationToken\` is not feature-gated, so \`default-features = false\` is sufficient and there is no transitive feature-unification dependency that could break.

File: src/twin/feature/fs_watcher.rs (hunk 16 above)
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs]: Fixed in 8952d6e: \`watch_dir_modified\` mask now includes \`MOVED_TO | MOVED_FROM\` alongside \`CREATE | DELETE\`, and the doc comment now says "additions, deletions and renames". Covered by existing \`file_created_via_rename\`, \`dir_modified_and_file_created_coexist_on_same_parent\`, and \`delete_event_does_not_fire_created_oneshot\` tests.

File: src/twin/system_info.rs (hunk 17 above)
Comment (JanZachmann) [reply src/twin/system_info.rs:218]: Documented in 8952d6e rather than code-fixed. \`/run/systemd/timesync/\` is created by systemd-timesyncd via \`RuntimeDirectory=\`, not by tmpfiles, so \`After=systemd-tmpfiles-setup.service\` does not guarantee it, and ods intentionally does not order \`After=time-sync.target\`. If the dir is absent the watch fails and startup aborts; systemd's restart policy (\`Restart=always\`) brings ods back, and since \`RuntimeDirectory\` is created when timesyncd *starts* (not when it syncs), the dir exists on the next attempt — self-healing within one \`RestartSec\` window. Added a comment at the watch site explaining this intended recovery path.

File: src/twin/mod.rs (hunk 18 above)
Comment (JanZachmann) [reply src/twin/mod.rs:280]: Fixed in 8952d6e: the disabled-feature no-reply path (FsEvent/Tick) now logs at \`debug!\`; the reply path (direct method / web service) keeps \`info!\` and still delivers the disabled reason.

File: src/twin/mod.rs (hunk 19 above)
Comment (JanZachmann) [reply src/twin/mod.rs:410]: Fixed in 8952d6e: added a comment documenting that a feature whose stream fails to build is logged and skipped rather than fatal, and that setup errors which must abort the service (e.g. inotify init) are already surfaced via \`FsWatcher\` in \`Twin::new()\`.

File: src/twin/feature/command.rs (hunk 20 above)
Comment (JanZachmann) [reply src/twin/feature/command.rs]: Fixed in 8952d6e: reworded to "Capacity 1 is backpressure, not an ordering guarantee" and noted that a slow consumer blocking \`tx.send().await\` prevents the \`cancel.cancelled()\` arm from being polled, lagging shutdown by up to one drain cycle.

File: src/twin/feature/command.rs (hunk 21 above)
Comment (JanZachmann) [reply src/twin/feature/command.rs:258]: Fixed in 8952d6e: added \`debounced_command_stream_source_ends_after_events_test\`, which sends an event, drops the source sender, and asserts the output stream ends (\`Ok(None)\`) rather than stalling — exercising the source-stream-ended recovery path.

File: src/twin/feature/fs_watcher.rs (hunk 22 above)
Comment (JanZachmann) [reply src/twin/feature/fs_watcher.rs:356]: Already covered in 2377e9a by \`handle_event_q_overflow_breaks_loop\`: it synthesizes a \`Q_OVERFLOW\` event, calls \`handle_inotify_event\`, asserts the loop breaks, and asserts no debounce deadline was armed. Companion tests \`handle_event_terminal_error_breaks_loop\` and \`handle_event_stream_end_breaks_loop\` cover the other exit paths.

**Review by JoergZeidler — APPROVED** [review]

File: src/twin/consent.rs [hunk 23]
```diff
@@ -119,6 +110,16 @@ impl DeviceUpdateConsent {
     const USER_CONSENT_VERSION: u8 = 1;
     const ID: &'static str = "device_update_consent";
 
+    pub fn new(fs_watcher: &mut FsWatcher) -> Result<Self> {
+        for path in [request_consent_path!(), history_consent_path!()] {
+            fs_watcher.watch_file_modified::<DeviceUpdateConsent>(&path)?;
+        }
```
Comment (mlilien) [inline src/twin/consent.rs:127]: claude:
watch_file_modified (consent) uses CLOSE_WRITE on the file inode; atomic rename-replace is silently missed.

- consent.rs::new watches request_consent.json + history_consent.json via watch_file_modified → WatchMask::CLOSE_WRITE on the file inode.                                                                                                                                                                             │
     │ - The PR's own watch_dir_modified doc states config tools/swupdate write via atomic rename-swap, which is why the dir watch includes MOVED_TO/MOVED_FROM. But the consent watch is on the file inode: an atomic replace (write tmp + rename over target) produces no CLOSE_WRITE on the watched inode, and leaves the │
     │ watch bound to the now-replaced inode — every later update is missed too.                                                                                                                                                                                                                                             │
     │ - request_consent.json is the swupdate-side consent trigger (written externally, not by ods to_json_file, which writes in-place). If the OS writes it via rename, the consent feature stops reacting — a functional regression for the feature's primary purpose.                                                     │
     │ - Needs verification: confirm how request_consent.json/history_consent.json are written by meta-omnect. If atomic-rename, watch the parent consent dir (mirroring watch_dir_modified) instead of the file inode.

File: src/twin/mod.rs [hunk 24]
```diff
@@ -71,18 +74,22 @@ impl Twin {
             - init features first
             - start with SystemInfo in order to log useful infos asap
         */
+        #[cfg(not(feature = "mock"))]
+        let mut fs_watcher = FsWatcher::new()?;
```
Comment (mlilien) [inline src/twin/mod.rs:78]: claude: 

Watch registration in Feature::new() is not gated by is_enabled()

connect_twin, connect_web_service, and command_request_stream are all is_enabled()-gated; the new in-new() watch setup is not. Twin::new() constructs every feature unconditionally and each constructor installs its kernel watches.                                                                               │
     │ - Consequence: with SUPPRESS_DEVICE_UPDATE_USER_CONSENT=true, DeviceUpdateConsent::new still calls watch_file_modified, which inotify_add_watch-fails (ENOENT) if the consent files are absent → Twin::new() bails → systemd restart loop. Old code (default()) read/watched nothing when suppressed.                 │
     │ - factory_reset is not affected (the dir is already a hard precondition of factory_reset_keys() in new()). The risk is specific to consent, where new() previously had no file precondition.                                                                                                                          │
     │ - In practice the consent files likely ship in the image (enabled connect_twin already requires them), so this may be latent — but the asymmetry is real. Fix: gate watch registration on is_enabled(), or document the unconditional file-existence precondition.

File: src/twin/consent.rs (hunk 23 above)
Comment (JanZachmann) [reply src/twin/consent.rs:127]: Verified the writers; this is not a regression. Both writers of \`request_consent.json\`/\`history_consent.json\` write **in place** on the same inode, so \`CLOSE_WRITE\` on the watched inode fires:
- swupdate side (iot-hub-device-update \`swupdate_consent_handler.cpp\`) uses parson \`json_serialize_to_file_pretty\`, which is \`fopen(path, "w")\` (truncate-in-place), no temp+rename.
- ods writes via \`common::to_json_file\` → \`OpenOptions::new().write(true).truncate(true).open(path)\` — also in place.

Since neither writer does atomic rename-replace, the file-inode \`CLOSE_WRITE\` watch is correct here and no events are missed. The \`MOVED_TO\`/\`MOVED_FROM\` mask on \`watch_dir_modified\` is for directory watches where external tools do rename-swap; it does not apply to these file watches.

File: src/twin/mod.rs (hunk 24 above)
Comment (JanZachmann) [reply src/twin/mod.rs:78]: Fixed in f7620c2: \`DeviceUpdateConsent::new\` now skips \`watch_file_modified\` registration when \`SUPPRESS_DEVICE_UPDATE_USER_CONSENT=true\`. You are right that the dispatch-time gating in 7c7fdb9 (\`handle_request\`) cannot cover this — it runs after construction, whereas the \`inotify_add_watch\` ENOENT happens during \`Twin::new\`. This is a construction-time precondition rather than dispatch gating, so it is consistent with keeping dispatch centralized.

Note on verification: this path is not unit-testable here because \`FsWatcher\` is a noop in mock/test builds, so there is no real \`add_watch\` to fail — the fix is by the env gate + reasoning, not a new test. On standard images the consent files ship via the iot-hub-device-update recipe, so the abort only occurs on images built without ADU while consent is suppressed.

**Review by mlilien — APPROVED** [review]
